pkg/scheduler/internal/cache/cache.go

```diff
@@ -124,22 +201,43 @@ func (cache *schedulerCache) Snapshot() *Snapshot {
 	}
 }
 
-func (cache *schedulerCache) UpdateNodeNameToInfoMap(nodeNameToInfo map[string]*schedulernodeinfo.NodeInfo) error {
+// SnapshotNodeInfo takes a snapshot of cached NodeInfo map. This is called at
+// beginning of every scheduling cycle.
+// This function tracks generation number of NodeInfo and updates only the
+// entries of an existing snapshot that have changed after the snapshot was taken.
+func (cache *schedulerCache) SnapshotNodeInfo(nodeSnapshot *NodeInfoSnapshot) error {
 	cache.mu.Lock()
 	defer cache.mu.Unlock()
+	balancedVolumesEnabled := utilfeature.DefaultFeatureGate.Enabled(features.BalanceAttachedNodeVolumes)
+
+	// Get the last generation of the the snapshot.
+	snapshotGeneration := nodeSnapshot.Generation
 
-	for name, info := range cache.nodes {
-		if utilfeature.DefaultFeatureGate.Enabled(features.BalanceAttachedNodeVolumes) && info.TransientInfo != nil {
+	// Start from the head of the NodeInfo doubly linked list and update snapshot
+	// of NodeInfos updated after the last snapshot.
+	for node := cache.headNode; node != nil; node = node.next {
+		if node.info.GetGeneration() <= snapshotGeneration {
+			// all the nodes are updated before the existing snapshot. We are done.
+			break
+		}
+		if balancedVolumesEnabled && node.info.TransientInfo != nil {
 			// Transient scheduler info is reset here.
-			info.TransientInfo.ResetTransientSchedulerInfo()
+			node.info.TransientInfo.ResetTransientSchedulerInfo()
```

Comment:
I know you're not changing this logic, but I don't like that it is here and I will look into it later.